File: Combinations/Combinations.java
public class Solution {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        combine(res, new ArrayList<Integer>(), 1, n, k);
        return res;
    }
    private void combine(List<List<Integer>> res, List<Integer> list, int start, int n, int k) {
        if (list.size() == k) {
            res.add(new ArrayList<Integer>(list));
            return;
        }
        for (int i = start; i <= n; ++i) {
            list.add(i);
            combine(res, list, i + 1, n, k);
            list.remove(list.size() - 1);
        }
    }
}

File: Combinations/Combinations.cpp
class Solution {
public:
    vector<vector<int> > combine(int n, int k) {
        vector<vector<int> > res;
        vector<int> list;
        combine(res, list, 1, n, k);
        return res;
    }
    void combine(vector<vector<int> > &res, vector<int> &list, int start, int n, int k) {
        if (list.size() == k) {
            res.push_back(vector<int>(list));
            return;
        }
        for (int i = start; i <= n; ++i) {
            list.push_back(i);
            combine(res, list, i + 1, n, k);
            list.pop_back();
        }
    }
    
};